Lib/pydoc.py

```diff
@@ -384,21 +385,34 @@ def ispackage(path):
     return False
 
 def source_synopsis(file):
-    line = file.readline()
-    while line[:1] == '#' or not line.strip():
-        line = file.readline()
-        if not line: break
-    line = line.strip()
-    if line[:4] == 'r"""': line = line[1:]
-    if line[:3] == '"""':
-        line = line[3:]
-        if line[-1:] == '\\': line = line[:-1]
-        while not line.strip():
-            line = file.readline()
-            if not line: break
-        result = line.split('"""')[0].strip()
-    else: result = None
-    return result
+    """Return the one-line summary of a file object, if present"""
+
+    if hasattr(file, 'buffer'):
+        file = file.buffer
+    if isinstance(file, io.TextIOBase):
+        try:
+            source = file.read()
+            if isinstance(source, bytes):
+                source = source.decode('utf-8')
+        except UnicodeEncodeError:
+            source = file.read().decode('latin-1')
+    else:
+        # Binary file
+        try:
+            source = tokenize.untokenize(tokenize.tokenize(file.readline))
+        except (SyntaxError, tokenize.TokenError, UnicodeDecodeError, ValueError):
+            return None
+
+    try:
+        tree = ast.parse(source)
+        if (tree.body and isinstance(tree.body[0], ast.Expr) and
+            isinstance(tree.body[0].value, ast.Constant) and
+            isinstance(tree.body[0].value.value, str)):
```

Comment:
Minor: use different indentations between if clause and suite
```suggestion
                isinstance(tree.body[0].value, ast.Constant) and
                isinstance(tree.body[0].value.value, str)):
```